modules/internal/BytesStringCommon.chpl

```diff
@@ -719,27 +722,95 @@ module BytesStringCommon {
         const localNeedle = needle.localize();
         const needleLen = localNeedle.buffLen;
 
-        // i *is not* an index into anything, it is the order of the element
-        // of view we are searching from.
-        const numPossible = xLen - nLen + 1;
-        const searchSpace = if fromLeft
-            then 0..#(numPossible)
-            else 0..#(numPossible) by -1;
-        for i in searchSpace {
-          // j *is* the index into the localNeedle's buffer
-          for j in 0..#nLen {
-            const idx = view.orderToIndex(i+j); // 0s based idx
-            if x.buff[idx] != localNeedle.buff[j] then break;
-
-            if j == nLen-1 {
-              if count {
-                localRet += 1;
-              } else { // find
-                localRet = view.orderToIndex(i);
+        const numChar = 256;
+
+        if nLen <= 10 || xLen <= 100 {
+          // i *is not* an index into anything, it is the order of the element
```

Comment:
As both of these implementations are fairly long, I recommend we have other helpers that implements these algorithms and that this function should call them. Off the top of my head, I would probably make them

```chapel
private proc doSearchNoEncBasic(...) {...}
private proc doSearchNoEncBMH(...) {...}
```

And when you do that, could you make sure that those helpers have some header comments describing what they do. You don't need to describe the full algorithm or anything, but the high-level distinction should be clear. i.e. the "basic" one does a straightforward substring search that's faster for shorter strings, and the other is implementing "Boyer-Moore-Horspool" algorithm and a link to a description of the algorithm.